File: apps/relay-server/src/prompt/generator-v3.ts
import type { Language, SessionMode, CollectedData } from '../types.js';
import {
  politenessRules,
  culturalRules,
  firstMessageTemplates,
  firstMessageAgentTemplates,
  termExplanationRules,
} from './templates.js';

interface SessionAPromptParams {
  mode: SessionMode;
  sourceLanguage: Language;
  targetLanguage: Language;
  collectedData?: CollectedData;
}

interface SessionBPromptParams {
  sourceLanguage: Language;
  targetLanguage: Language;
}

/**
 * Generate Session A system prompt based on mode and languages.
 */
export function generateSessionAPrompt(params: SessionAPromptParams): string {
  const { mode, sourceLanguage, targetLanguage, collectedData } = params;

  const langKey = `${sourceLanguage}->${targetLanguage}` as `${Language}->${Language}`;
  const politeness = politenessRules[langKey] ?? '';
  const cultural = culturalRules[langKey] ?? '';

  if (mode === 'relay') {
    return generateRelayModePrompt(sourceLanguage, targetLanguage, politeness, cultural, collectedData);
  }

  return generateAgentModePrompt(targetLanguage, politeness, collectedData);
}

function generateRelayModePrompt(
  sourceLanguage: Language,
  targetLanguage: Language,
  politeness: string,
  cultural: string,
  collectedData?: CollectedData,
): string {
  const sourceName = sourceLanguage === 'en' ? 'English' : 'Korean';
  const targetName = targetLanguage === 'en' ? 'English' : 'Korean';

  const contextBlock = collectedData
    ? `
## Context
You are making a phone call to ${collectedData.targetName} on behalf of the user.
Purpose: ${collectedData.scenarioType} - ${collectedData.service}
Customer Name: ${collectedData.customerName}`
    : '';

  const firstMessage = firstMessageTemplates[targetLanguage];

  return `You are a real-time phone translator.
You translate the user's speech from ${sourceName} to ${targetName}.

## Core Rules
1. Translate ONLY what the user says. Do NOT add your own words.
2. ${politeness}
3. Output ONLY the direct translation. No commentary, no suggestions.
4. Adapt cultural expressions naturally:
   ${cultural}
5. For place names, use the local name (e.g., "Gangnam Station" → "강남역").
6. For proper nouns without local equivalents, transliterate them.
${contextBlock}

## First Message (AI Disclosure — auto-generated)
When instructed with [SYSTEM], say exactly the given message.
Default first message: "${firstMessage}"

## CRITICAL: You are a TRANSLATOR, not a conversationalist.
- Do NOT answer questions from the recipient on your own.
- Do NOT make decisions on behalf of the user.
- If the recipient asks something, translate it to the user and wait.`;
}

function generateAgentModePrompt(
  targetLanguage: Language,
  politeness: string,
  collectedData?: CollectedData,
): string {
  const targetName = targetLanguage === 'en' ? 'English' : 'Korean';

  const collectedBlock = collectedData
    ? `
## Collected Information
${JSON.stringify(collectedData.details, null, 2)}

## Task
${collectedData.scenarioType}: ${collectedData.service}
Target: ${collectedData.targetName} (${collectedData.details?.phone ?? 'N/A'})`
    : '';

  const fillerPhrase = targetLanguage === 'ko'
    ? '잠시만요, 확인하고 말씀드릴게요'
    : "One moment please, let me check on that";

  const firstMessage = collectedData
    ? firstMessageAgentTemplates[targetLanguage].replace('{{service}}', collectedData.service)
    : firstMessageAgentTemplates[targetLanguage].replace('{{service}}', 'your service');

  return `You are an AI phone assistant making a call on behalf of a user who cannot speak directly.

## Core Rules
1. Use polite ${targetName} speech at all times. ${politeness}
2. Complete the task based on the collected information below.
3. If the recipient asks something you don't have the answer to,
   say "${fillerPhrase}" and wait for the user's text input.
4. Keep responses concise and natural, like a real phone conversation.
${collectedBlock}

## Conversation Strategy
1. Greet and state the purpose.
2. Provide collected information as needed.
3. Confirm details when asked.
4. Thank and close when task is complete.

## First Message
Default greeting: "${firstMessage}"

## When You Don't Know the Answer
- Say the filler phrase: "${fillerPhrase}"
- Wait for text input from the user via conversation.item.create.
- Relay the user's text response naturally in speech.`;
}

/**
 * Generate Session B system prompt.
 */
export function generateSessionBPrompt(params: SessionBPromptParams): string {
  const { sourceLanguage, targetLanguage } = params;

  const sourceName = sourceLanguage === 'en' ? 'English' : 'Korean';
  const targetName = targetLanguage === 'en' ? 'English' : 'Korean';
  const langKey = `${targetLanguage}->${sourceLanguage}` as `${Language}->${Language}`;
  const termRules = termExplanationRules[langKey] ?? '';

  // KR→KR scenario: no translation needed, just STT
  if (sourceLanguage === targetLanguage) {
    return `You are a real-time speech-to-text transcriber.
You transcribe the recipient's speech in ${targetName} accurately.

## Core Rules
1. Output ONLY the direct transcription. No commentary.
2. Preserve the speaker's intent and tone.
3. Do NOT add your own opinions.
4. Do NOT respond to the recipient.`;
  }

  return `You are a real-time translator.
You translate the recipient's speech from ${targetName} to ${sourceName}.

## Core Rules
1. Translate what the recipient says into natural ${sourceName}.
2. Output ONLY the direct translation.
3. Preserve the speaker's intent, tone, and urgency.
4. For culture-specific terms, add brief context in parentheses:
   ${termRules}
5. For time/currency references, convert to the user's context.

## Do NOT:
- Add your own opinions or suggestions.
- Summarize or skip parts of the conversation.
- Respond to the recipient (you are only translating).`;
}

File: apps/relay-server/src/prompt/generator_v3.py
"""v3 System Prompt 생성기 — Relay/Agent 모드별, 양방향 언어.

PRD 8.1 기반. 수집된 데이터와 언어 설정으로 동적 프롬프트를 생성한다.
"""

import json
import logging
from typing import Any

from src.prompt.templates import (
    CULTURAL_ADAPTATION_RULES,
    POLITENESS_RULES,
    SESSION_A_AGENT_TEMPLATE,
    SESSION_A_RELAY_TEMPLATE,
    SESSION_B_TEMPLATE,
    TERM_EXPLANATION_RULES,
)
from src.types import CallMode

logger = logging.getLogger(__name__)


def generate_session_a_prompt(
    mode: CallMode,
    source_language: str,
    target_language: str,
    collected_data: dict[str, Any] | None = None,
) -> str:
    """Session A (User→수신자) 프롬프트를 생성한다."""
    lang_pair = (source_language, target_language)

    if mode == CallMode.RELAY:
        prompt = SESSION_A_RELAY_TEMPLATE.format(
            source_language=_lang_name(source_language),
            target_language=_lang_name(target_language),
            politeness_rules=POLITENESS_RULES.get(lang_pair, "Use polite speech."),
            cultural_adaptation_rules=CULTURAL_ADAPTATION_RULES.get(
                lang_pair, "Adapt naturally."
            ),
            target_name=_get(collected_data, "target_name", "the recipient"),
            scenario_type=_get(collected_data, "scenario_type", "general inquiry"),
            service=_get(collected_data, "service", ""),
            customer_name=_get(collected_data, "customer_name", "the customer"),
        )
    else:
        prompt = SESSION_A_AGENT_TEMPLATE.format(
            target_language=_lang_name(target_language),
            collected_data=json.dumps(collected_data or {}, ensure_ascii=False, indent=2),
            scenario_type=_get(collected_data, "scenario_type", "general inquiry"),
            service=_get(collected_data, "service", ""),
            target_name=_get(collected_data, "target_name", "the recipient"),
            target_phone=_get(collected_data, "target_phone", ""),
        )

    logger.info(
        "Generated Session A prompt (mode=%s, %s→%s, %d chars)",
        mode.value,
        source_language,
        target_language,
        len(prompt),
    )
    return prompt


def generate_session_b_prompt(
    source_language: str,
    target_language: str,
) -> str:
    """Session B (수신자→User) 프롬프트를 생성한다."""
    lang_pair = (target_language, source_language)

    prompt = SESSION_B_TEMPLATE.format(
        source_language=_lang_name(source_language),
        target_language=_lang_name(target_language),
        term_explanation_rules=TERM_EXPLANATION_RULES.get(
            lang_pair, "Add brief context for culture-specific terms."
        ),
    )

    logger.info(
        "Generated Session B prompt (%s→%s, %d chars)",
        target_language,
        source_language,
        len(prompt),
    )
    return prompt


def _lang_name(code: str) -> str:
    names = {"en": "English", "ko": "Korean", "ja": "Japanese", "zh": "Chinese"}
    return names.get(code, code)


def _get(data: dict | None, key: str, default: str = "") -> str:
    if data and key in data:
        return str(data[key])
    return default
